### ISSUE TITLE
Proposal/RFC: Wadm tidying and productionizing

### ISSUE THREAD

**Issue by thomastaylor312** [wasmCloud/wadm#40]
## Overview

We started Wadm a long time ago, but have mostly let it sit since then as we've worked on polishing the host. However, the time has come to get this all working and productionized. This document is a proposal for rewriting and releasing wadm as a fully supported and featured part of the wasmCloud ecosystem, [complete with waterslide!](https://getyarn.io/yarn-clip/1254a73a-099a-4c2d-a340-2bee36aeb744)

## Goals and Non-goals

This section covers the goals and non-goals of this scope of work. Please note that some of the items in non-goals are possible future work for wadm, _but are not in scope_ for this work

### Goals

- Make it easy to run a wasmCloud application in a lattice, including redistributing and running an application as hosts join or leave a lattice
- Give users a declarative way to run their applications
- wasmCloud operators should have a clear guide on how to run and deploy wadm
    - This also includes new users who just want to run it, meaning this should be included in `wash up`
- Be the canonical scheduling implementation for wasmCloud without precluding the use of other custom built schedulers
- Be a project that people want to contribute to
    - I know this one sounds all vision-y and corporate-y, but it is a key point. If we want to make something robust that will serve the needs of many people, we need contributors
- Have planned extension points for custom scalers (implementation is not required)

### Non-goals

- Make wadm a full control plane for all things
- HTTP API for wadm
- Changing the scope of features that already exist in the current version of wadm

### Language choice

For this work, I propose we rewrite Wadm in Rust. This decision was made for the following reasons, in order of importance. As part of making this decision, two other languages (Elixir and Go) were considered. Reasons for rejecting are described in the last 2 sections

#### Need for contributors

Schedulers and application lifecycle management are topics that many people in the cloud native space have deep knowledge of. If we are going to be writing something that does those things for wasmCloud, then we need as many eyes on it as possible. Based on current metrics of wasmCloud repos, we have very few contributors to our Elixir code and a lot more to our Rust repos. Other projects in, or consumed by, the wasm ecosystem are in Rust and also have higher numbers of contributors. Go would have also been an excellent choice here, but the other reasons listed here precluded it. We also have multiple contributors in the wasmCloud community right know who already know Rust.

The tl;dr is that we need contributors to be successful and the current language does not attract enough people.

#### Static Typing and Generics

One problem we've run into consistently in our elixir projects is issues with dynamic typing. Although this can be mitigated somewhat by tools like dialyzer, it requires programmer and maintainer discipline and still doesn't catch everything. Having a static type for each type of event that will drive a system like Wadm is critical for ensuring correct behavior and avoiding panics.

In addition to the need for static typing is the preference for having generics. In my experience with writing large applications for Kubernetes in both Rust and Go, a generic type system makes interacting with the API much easier. There is less generated code and need for rolling your own type system as what happens in many large Go projects. Go has added generics, but its system is nowhere near as strong as other statically typed languages such as Rust.

#### Support for wasm and wasmCloud libraries

To support custom scalers, we will likely be supporting at least an actor based extension point and possibly a bare wasm module. Either way, most of the wasm libraries/runtimes out in the wild are written in Rust or have first-class bindings for Rust. Also, many of our wasmCloud libraries are written in Rust, which will allow for reuse.

#### Static binaries and small runtime size

This is the lowest priority reason why I am suggesting Rust, but it is still an important one. The current implementation requires bundling an Erlang VM along with the compiled Elixir. That means someone who runs Wadm as it currently stands will likely need to tune a VM. It is also larger, which leads to more space requirements on disk and longer download times.

Rust (and even Go moreso) has great support for static binaries and both run lighter than a VM without much additional tuning (if any).

#### Disadvantages of Rust

As with any tool choice, there are tradeoffs that occur. Below is a list of disadvantages I think will be most likely to cause friction

- Rust async
    - This isn't as bad of a problem as some people in the Rust community say, but we will likely need to implement things/workaround some of these rough edges
- Dealing with Rustls
- Rust still has a steeper learning curve than Go, so there will need to be some handholding on PRs from new contributors

#### Why not Elixir?

One of the biggest questions here is why not continue with Elixir. By far the biggest thing we are giving up is the code around the lattice observer. However, writing this in Rust gives us the advantage of creating something that we could eventually make bindings for in any language (this also helps enable the reusability described [below](#reusability-and-a-canonical-control-plane)), although that isn't a goal here. 

With that said, the previous sections cover in depth the advantage of using Rust over Elixir in this case

#### Why not Go?

In my comparisons, I was looking for languages that would fit the requirements above. Due to the overlap of languages used for wasm as well as languages familiar to those in the Cloud Native space, that whittled things down to Go and Rust. Go in many ways excels at many of these requirements. It is much more popular that Rust and Elixir (probably combined) and has great support for statically compiling binaries. Also, things like NATS are native to Go. 

It came down to a few main concerns of why Rust would be better:

- Generics + Code cleanliness
- Most of Wasm and wasmCloud are in Rust
- Familiarity and preference of core maintainers

To be clear, there are other smaller reasons, but those could be considered nitpicky.

### State machine vs event-driven "filtering"

One of the items I most thought about when drafting this was whether or not we should implement wadm as a true state machine. Given the simplicity of what it is trying to do, I propose we focus more on implementing an event-driven filtering approach. Essentially, a state machine approach is going to overkill for this stage of the project and the near future.

Loosely, I am calling these "Scalers" (name subject to change). Every scaler can take a list of events (that may or not be filtered) that returns a list of actions to take.

This does not mean we might iterate into a state machine style in the future (if you are curious, you can see [Krator](https://github.com/krator-rs/krator) for an example of how this could be done in code) or that a scaler implementation can't use a state machine. This _only_ means that for this first approach, we'll filter events into actions.

I have purposefully not gone into high levels of detail of what this looks like in code as it will probably be best just to try and see how this looks like as we begin to implement it. What we [currently have](https://github.com/wasmCloud/wadm/blob/53ca23ebf402ac5dd325b8a340ef956b1717fae4/wadm/lib/wadm/reconciler/command.ex#L7) in wadm is probably a good way of going about this (i.e. Scalers output commands)

### Scalers are commutative

One important point is that these "Scalers" should be [commutative](https://en.wikipedia.org/wiki/Commutative_property) (i.e. if a+b=c then b+a=c, the order of operations doesn't matter). That means when a manfest is sent to wadm, it can run through the list of supported Scalers in any order and it will return the same output.

### API is NATS-based

For this first production version, we will only be supporting a NATS API. This is because pretty much all wasmCloud tooling already uses NATS mostly transparently to an every day user. We can take advantage of that same tooling to keep things simple this time around. If we were to add an HTTP API right now, we'd have to figure out authn/z and figure out how we want to handle issuing tokens. So to keep it simple, we'll focus only on NATS to start.

One very important note here is that we _definitely do want an HTTP API_ in the future. We know that many people will want to integrate with or extend WADM and an HTTP API is the easiest way to do that. But not for this first go around (well, second, but you get my point)

### Data is stored in NATS KV

This is fairly self explanatory, but we want to store everything in one place now that NATS has KV support so we don't need any additional databases. Only the manifest data is stored in NATS. Lattice state is built up by querying all hosts on startup and then responding to events

### High availability

A key requirement is that wadm can be run in high availability scenarios, which at its most basic means that multiple copies can be running.

I propose that this be done with leader election. Only one wadm process will ever be performing actions. All processes can gather the state of the lattice for purposes of fast failover, but only one performs actions. This is the simplest way to gain basic HA support

### Custom scheduler support

This is purely here as a design note and is not required for completing the work, but based on experiences with tools like Kubernetes and Nomad, extending with a custom scheduler is a common ask for large deployments. In code, adding a scaler will be as simple as implementing a `Scaler` trait.

For most people however, I propose that custom "Scalers" be added via a wadm manifest. The application provider must have an actor that will implement a new `wasmcloud:wadm-scaler` interface, but can be as arbitrarily complex as desired. This manifest will have 2 special requirements

1. It must have an annotation like `wasmcloud.dev/scaler: <scaler-name>`
2. It is only allowed to use the built in Scaler types to avoid chicken and egg problems

Once again, this is not going to be implemented here, and will likely be another, smaller, proposal than this one

### Reusability and a canonical scheduler

One key point to stress here is that wadm is meant to be the canonical scheduler for wasmCloud. This means that it is the general purpose scheduler that most people use when running wasmCloud, but no one is forced to use it. You can choose not to use it at all, or to write your own entirely custom scheduler.

To that end, I propose we publish the key functionality as a Rust crate. Much of the functionality could be used in many other applications besides a scheduler, but it can also be used to build your own if so desired. Basically we want to avoid some of the problems of what occurred in Kubernetes where everything _must_ go through the built-in scheduler

## Basic roadmap of work

Whew, we made it to the actual work! As part of thinking through these ideas, I started [a branch](https://github.com/thomastaylor312/wadm/tree/feat/new_wadm) that has implemented some of the basic building blocks like streaming events and leader election. When we actually begin work, it will be against a new `wadm_0.4` branch in the main repo until we have completed work. Please note that these are a general roadmap, I didn't want to try and give minute details here. Below is the basic overview of needed work.

### Stage 1

All of this work can be worked on in parallel. This is a bit shorter because we are about 40-50% there with the branch I started work on

- Implement receiving and parsing wadm/OAM manifests, then storing them in NATS. 
- Build up lattice state from queries (reconciliation loop) and events

### Stage 2

This is a bit more difficult as these things must be worked on roughly in order. This work is more spike-like as it is spiking out the design of the Scaler

- Define the `Scaler` trait and implement the spreadscaler type (at least the number of replicas functionality). Scalers will need to handle manifests and state changes given by events (such as if a host stops). We want to start with implementing so we can see what kind of info is needed
- Create a work loop that takes incoming manifest changes and events and gives the list of scalers all it needs to process that event

### Stage 3

- Fully implement the spreadscaler type
- Implement linkdef handler
- Run e2e tests with multiple wasmCloud hosts
- Fully functional pipelines and alpha deploy candidate (crate, plain binary, and docker container)

### Stage 4

This is the "tying a bow on it" stage of work

- Add wadm support to `wash up` by default with an optional flag to not use it
- All flags and configuration documented
- Create deploy guides for getting started and "production level" deployments with multiple wadm's running with TLS against NATS. Should also show how to use the docker container.

**Comment by stevelr** [comment]
If wadm does need to make decisions as a leader, another trick may be to let jetstream do the leader election for you - it already does elections for nats servers ..

- Pair each wadm with a nats server in a nats server cluster and let the nats servers do their leader election in the JS domain. The rust nats client can tell you if your server is the leader - leader election done!

- For publishing leader decisions, like updates to global state, onto a durable stream, a node must check that it's a leader via the nats client, before publishing to the stream. If they are in a split-braine network and not the real leader, the nats client should indicate that there is no leader.

- Any subscriber to the stream always gets the correct current global state

**Comment by thomastaylor312** [comment]
@autodidaddict I really like the streams idea. One thing to follow up on is around what @stevelr was getting at. The only state we are storing are the submitted OAM manifests, but would the solution be to have a durable stream for the put and delete commands so that only one process updates state?

**Comment by autodidaddict** [comment]
You wouldn't need another stream for the put and delete commands since that's basically what NATS implements for KV under the hood. I alluded to it above but you can have the wadm processes listen to changes to the specs so that if they're caching anything internally, they can refresh it. You might also want to persist the last known observed state of a lattice based on the last processed event. This would allow a full wadm wipe and then when one comes back up, it immediately can recover it's last meaningful state. 

Of course, you can also issue a lattice-wide host ping to recover most of that observed state as well. Over the course of history, we've gone back and forth on whether it's better to have persisted observed state. If each wadm process is receiving events from the event stream _individually_, then you'll want the persisted state of the _observed lattice_ to also probably be in the KV store so that the next wadm to receive an event can react to the latest consistent state.

**Comment by thomastaylor312** [comment]
So would things change to basically put a wadm manifest directly into the KV bucket?

This actually made me realize something that applies to both the manifests and the events. There are two things that have to happen:

1. Every running wadm should receive a copy of the event/manifest in order to update internal caches/lattice state
2. Only _one_ wadm should generate commands

As far as I know a stream can't do that, correct? Would the solution be to create something like a push consumer that is used purely for updating internal caches and then actual _processing_ of an event would be done by the streams as described above?

**Comment by thomastaylor312** [comment]
I just saw your updated comment. Looks like you answered the question by saying that each process that receives an event should update that state into a KV bucket? So for lattice state, would we write that to the KV too?

**Comment by autodidaddict** [comment]
yeah, so there's some wiggle room here but I'm thinking we end up with:

* `wadm-events` - stream populated by publications to `wasmbus.evt.>`, processor pulls from this, generates new state, writes new state, publishes compensating action, acks message
* `wadm-commands` - stream populated by publishing compensation action requests, processor pulls from this, converts into lattice controller (et al) commands, ack to remove
* `wadm-observed` - a KV bucket with the key being a lattice ID and the value is the state as written by the last event processor

**Comment by autodidaddict** [comment]
this last KV bucket simplifies the architecture so that we don't have to manually synchronize the observed state across all running wadm instances - instead they pull the state relevant for the inbound event (which can be for many different lattices)

**Comment by stephenandary** [comment]
This almost sounds like a blackboard design pattern! https://en.wikipedia.org/wiki/Blackboard_(design_pattern), would that be a good tangent to draw here?

**Comment by autodidaddict** [comment]
@stephenandary I hadn't seen that one before. At a high level wadm is basically a `control loop`.  As state changes, it performs a reconciliation that determines a gap between the current state and the desired state. Then it issues commands that are designed to close that gap to make desired == reality. This loop continues every time the current state diverges from desired.

Kubernetes controllers are fairly well known examples of control loops, which some people also call `autonomous agent`s

Where NATS streams come in is to provide a durable way to guarantee responding to incoming events and guarantee that outbound commands are stored and acted upon by these autonomous agents.

**Comment by stephenandary** [comment]
@autodidaddict 

I am fairly familiar with k8s and infra though it has been some time since I have worked in a production environment now. I am very excited about WADM because what it can accomplish feels like the next stage of distributed systems due to NATS as you said, and when I search for other solutions, they don't seem great. 

What you are describing almost perfectly aligns with how a blackboard design pattern works, I learned about that pattern in a book called The Pragmatic Programmer, and it has fascinated me for some time. 

I will break it down quickly because I think it is an excellent design pattern for this situation, and I think what was discussed above fits the model: 

The blackboard model describes three main components of a non-deterministic solution:

- The Blackboard is structured global memory
- Knowledge Sources can be seen as highly specialized agents with their own representation. 
- The Control component selects, configures, and executes knowledge sources.


A Blackboard allows multiple Knowledge Sources (or agents) to communicate by reading and writing requests and information to a global data store (“the blackboard” - maybe NATS in this case). 

Each Knowledge Source is a participating agent that has expertise in their own field and has knowledge that is applicable to a part of the problem, i.e., the problem cannot be solved by an individual agent only. Knowledge Sources communicate strictly through a common blackboard whose contents are visible to all agents. When a problem is to be solved, candidate agents that can handle it are listed. 

A Control unit is responsible for selecting among the candidate knowledge sources to assign the task to one of them (I am not sure if this would be the same as a leader, but I don’t think it is). 

When we start to think about deploying applications across multiple lattices/clusters, it becomes very interesting to consider the possibility of having agents that have knowledge of data like network bandwidth, memory bandwidth, I/O bandwidth, call latency, etc that can communicate this information through to some blackboard to determine to scale, or even rightsize across the supercluster as a whole such that no individual lattices/clusters are seeing excessively high metrics in relation to their counterparts. 

To accomplish this Knowledge Sources contain thresholds by which they become activated and can update the blackboard and await commands to do some execution. There are then higher-level thresholds in the Blackboard, which when met can trigger an event whereby it can list activated knowledge sources capable of solving the problem, and the Control can select the best candidate from the available list within the Blackboard to do so.

**Comment by autodidaddict** [comment]
:brain: This is fascinating. Right now wadm really only has one knowledge source - the state of an observed lattice. But as you point out, it could include things like network measurements, CPU activity from hosts, latency, and so forth.

**Comment by autodidaddict** [comment]
It contains a state representation of the current lattice. So, all of the claims, link definitions, actors, providers, hosts, and invocation counts. Wadm currently uses the model in [lattice-observer ](https://github.com/wasmCloud/lattice-observer/tree/main/lib/lattice_observer/observed) for its observation state.

**Comment by thomastaylor312** [comment]
FWIW, I personally think that wadm could be the way we expose current stats/metrics in the future (a la control plane), but haven't defined anything specific yet

**Comment by stephenandary** [comment]
I feel this loops back to the Blackboard pattern again, specifically knowledge sources. In the current WADM state, there is already a knowledge source via lattice observer. In K8s that may just be some sidecar that sends telemetry back to WADM. I feel that the issue isn't so much not being able to get telemetry, it is how. NATS clients can communicate via several different protocols though, right?

**Comment by thomastaylor312** [comment]
@autodidaddict I thought about maybe making metrics be exported on a specific topic!

**Comment by thomastaylor312** [comment]
Ok, back to topic now. I had an offline convo to discuss some more with @autodidaddict about his idea for using streaming. We came up with what I feel like is a good plan to move forward and mitigate any risk of too much complexity.

At the beginning of stage 1, we propose that we spike out a simple implementation of the stream plan as suggested in https://github.com/wasmCloud/wadm/issues/40#issuecomment-1440686958. It doesn't have to actually do any work, but we want to test receiving events from a stream and emitting commands (and updating state in the kv store). Then make sure the command is received and compensatory actions are done.

If for some reason that ends up being too complex or has other problems, we will fall back to the original leader election plan with two changes:

- We will still put all events into a durable stream for consumption to make sure nothing is missed
- State will not be in memory, but persisted to nats KV

This would allow us to more easily evolve the architecture to something more complex in the future. It is also important to note that leader election should be ok in the meantime since our only chance of a network partition is with NATS and if NATS is messed up, so is the rest of the lattice.

**Comment by stephenandary** [comment]
How will this streaming method handle multiple compensatory event emmissions from various observers or hosts if they are correlated events? Can a process be aware of other events when producing compensatory action? Will it verify if that event is still valid?

**Comment by stephenandary** [comment]
Is this the slice?

```mermaid
flowchart TD
    A[lattice_observer - knowledge source] -->| NATS: wadm_event - blackboard | B(wadm_process - control)
    B --> | NATS: wadm_action - blackboard | A
 ```

**Comment by autodidaddict** [comment]
Correlation is actually pretty "easy", given that everything emitted by the wasmCloud host is a cloud event with a unique ID. that can be used both to help ensure idempotency as well as correlate corrective actions taken by the controller in response to a given eent.

**Comment by autodidaddict** [comment]
I think we're on the same page here, the autonomous control loop within wadm is the thing that makes decisions

**Comment by stephenandary** [comment]
Ok awesome, and that is because the NATS stream is not persistent global memory?

**Comment by autodidaddict** [comment]
the NATS stream is durable and replicated and managed across a NATS cluster by the NATS server according to however we configure the stream(s) - e.g. eviction policy (ttl), number of replicas, etc. It's not global in-memory cache.

**Comment by autodidaddict** [comment]
NATS streams can be used in place of Kafka in architecture diagrams. Nobody's suggesting the blackboard be in-memory.

re: nomad, wadm's responsibilities including scheduling providers and actors. It's not currently in scope to manage the starting and stopping of wasmCloud hosts.

**Comment by stephenandary** [comment]
That is good to know. I was planning to use terraform for setting up
wasmcloud hosts if it is no longer necessary to have nomad or k8s manage
some containers or pods for applications to run in. Seems more simple that
way, maybe that won't work though, but I don't see why using wasm shouldnt
enable one to remove the need for pods and containers on a host completely.
Then it would be a matter of adding knowledge sources for hosts through the
wadm chart if that becomes possible. Maybe I am imagining things
incorrectly though.

**Comment by thomastaylor312** [comment]
@stephenandary I think this convo may have gotten into the weeds. What is important here is the high level view: 

- Wadm will ensure that the application manifests it has been given are properly spun up and running with their given requirements at all times (e.g. "I want 3 copies of X actor running on hosts with matching labels with 2 copies of Y provider spread across my regions with Z linkdefs"). 
- Wadm is not in charge of managing hosts nor has it ever had a concept of using pods or containers to run things. Containers/pods are a supported way to run a host or wadm, but not a requirement.
- Hopefully in the future, wadm can be the source of truth for all lattice state. However, this is not the primary goal of the first version

That first, simple point is the whole goal of the wadm project and this RFC is purely for why we are rewriting in Rust, and some basic design principles. I'm not trying to follow some specific "blackboard" pattern or anything else per se, just making sure you can run this in production and have work done once. Since that initial work is now front loaded, if there are any concerns around how that is designed/functioning, that conversation can be had in concrete terms around actual code.

Does that help clarify things?

**Comment by thomastaylor312** [comment]
Now that we have https://github.com/orgs/wasmCloud/projects/8/views/1, I am going to close this RFC
